syntax = "proto3";
package LabNetProt.Client;
import "LabNet.proto";

// IO board messages
message IoBoardInit {
}

message IoBoardInitDigitalIn {
	uint32 pin = 1;
	enum Resistor {
		OFF = 0;
		PULL_DOWN = 1;
		PULL_UP = 2;
	}
	Resistor resistor_state = 2;
	bool is_inverted = 3;
}

message IoBoardInitDigitalOut {
	uint32 pin = 1;
	bool is_inverted = 2;
}
// IO board messages


// RFID board messages
message RfidBoardInit {
	uint32 antenna_phase1 = 1;
	uint32 antenna_phase2 = 2;
	uint32 phase_duration = 3;
	bool inverted = 4;
}

message RfidBoardSetPhaseMatrix {
	uint32 antenna_phase1 = 1;
	uint32 antenna_phase2 = 2;
	uint32 phase_duration = 3;
}
	
message RfidBoardSetSignalInversion {
	bool inverted = 1;
}
// RFID board messages


// UART messages
message UartInit {
	Uarts port = 1;
	uint32 baud = 2;
}

message UartWriteData {
	Uarts port = 1;
	bytes data = 2;
}
// UART messages


// Gpio with WiringPi
message GpioWiringPiInit {
}

message GpioWiringPiInitDigitalIn {
	uint32 pin = 1;
	enum Resistor {
		OFF = 0;
		PULL_DOWN = 1;
		PULL_UP = 2;
	}
	Resistor resistor_state = 2;
	bool is_inverted = 3;
}

message GpioWiringPiInitDigitalOut {
	uint32 pin = 1;
	bool is_inverted = 2;
}
// Gpio with WiringPi

// digital out messages
message DigitalOutSet {
	PinId id = 1;
	bool state = 2;
}

message DigitalOutPulse {
	PinId id = 1;
	uint32 high_duration = 2;
	uint32 low_duration = 3;
	uint32 pulses = 4;
}
// digital out messages


// digital out loop
message StartDigitalOutLoop {
	message DigitalOutputParameter {
		PinId id = 1;
		uint32 offset = 2;
		uint32 duration = 3;
	}
	string loop_name = 1;
	uint32 loop_pause = 2;
	repeated DigitalOutputParameter digital_outputs = 3;
}

message StopDigitalOutLoop {
	string loop_name = 1;
}
// digital out loop


// general messages
message LabNetResetRequest {
}

message LabNetIdRequest {
}
// general messages

message ClientWrappedMessage {
	oneof client_message {
		LabNetResetRequest reset_request = 1;
		LabNetIdRequest id_request = 2;
		IoBoardInit io_board_init = 3;
		IoBoardInitDigitalIn io_board_init_digital_in = 4;
		IoBoardInitDigitalOut io_board_init_digital_out = 5;
		RfidBoardInit rfid_board_init = 6;
		RfidBoardSetPhaseMatrix rfid_board_set_phase_matrix = 7;
		RfidBoardSetSignalInversion rfid_board_set_signal_inversion = 8;
		UartInit uart_init = 9;
		UartWriteData uart_write_data = 10;
		DigitalOutSet digital_out_set = 11;
		DigitalOutPulse digital_out_pulse = 12;
		StartDigitalOutLoop start_digital_out_loop = 13;
		StopDigitalOutLoop stop_digital_out_loop = 14;
		GpioWiringPiInit gpio_wiring_pi_init = 15;
		GpioWiringPiInitDigitalIn gpio_wiring_pi_init_digital_in = 16;
		GpioWiringPiInitDigitalOut gpio_wiring_pi_init_digital_out = 17;
	}
}